### DISCUSSION TITLE
How to stop inputs losing focus in lists?

### DISCUSSION THREAD

**Discussion by dominictobias** [solidjs/solid#1589] [Q&A] [answered]
Outside of a For loop if I update an input (via `onInput` since I need to update the UI on input change not blur) the input element doesn't re-render:

https://playground.solidjs.com/anonymous/c3d7da21-6eea-473e-b680-d83992545217

However inside a for loop it re-renders the input on every keypress, thus loosing focus:

https://playground.solidjs.com/anonymous/33badcbc-2b19-4266-a82d-35459d9f62d1

How can I stop it from re-rendering the whole element and losing focus?

**Comment by NawfelBgh** [comment, marked as answer]
The problem you are facing has to do with how the `For` component works.

Both the `For` and `Index` component create a rendering effect per object from their input array. The difference between the two components is that when the input array is updated:

- `For` reruns the effect for every object that moved (changed index) inside the array, creates a new rendering effect for newly added objects and destroys the effects of objects that have been deleted
- where as `Index` reruns the effect for each index where the object has changed, creates a new rendering effect for newly added indices and destroys the effects for indices that no longer exist in the array

That is why `For`'s closure argument takes an item value and index signal where as `Index`'s closure argument takes an item signal and an integer value for index. 

In your example, when the array item value changes, `For` considers it deleted from the array so it destroys its rendering effect, and it sees a new object added in it's place for which it creates a new renderering effect.

Here is your example modified to use the `Index` component which solves the issue: https://playground.solidjs.com/anonymous/fc4d0762-62d2-45f5-b157-87d9120d5904

**Reply by dominictobias** [reply]
Oh great I wasn't aware of Index or the behavior thanks!
